# What are Shapley Values?

Commonly reffered to as SHAP (SHapley Additive exPlanations) values, they provide a way of understanding the true "impact" of a variable/feature in a specific task and how it may impact the outcome of said task.

It's as if we try to measure the degree of *"importance"* a single contribution to a specific task has.

They are based on **Game Theory**, and assign a number to each feature in a model. Features with positive SHAP values positively impact the prediction, while those with negative values have a negative impact.

### Why am I writing about them? 

Here is the **importance** 🥁 and usefulness of Shapley Values:

* Shapley values provide a robust and mathematically sound method to explain individual predictions of complex models, such as ensemble methods (XGBoost, Random Forest) or neural networks. They help break down the contribution of each feature to a specific prediction, making black-box models more transparent.

* Understanding why a model made a certain decision allows you to explain predictions in a way that is consistent and fair, helping users understand and trust the model’s output.

* Shapley values can highlight if certain features are disproportionately influencing decisions, which helps identify and mitigate biases in the model.

* They give insight into feature importance on a granular level by considering all possible combinations of features. 

* In industries like finance or healthcare, regulatory requirements demand model transparency. Shapley values can help meet these requirements by providing explanations for predictions that are mathematically sound and easy to interpret.

* By understanding how features contribute to predictions, you can diagnose and fine-tune models more effectively. Shapley values provide insight into potential areas for improvement, such as feature selection, handling outliers, or addressing multicollinearity.

# SHAP values: definition and properties

Shapley Values are computed using the following expression:

$$
\phi_j(\text{val}) = \sum_{S \subseteq \{1,\dots,p\} \setminus \{j\}} \frac{|S|! (p - |S| - 1)!}{p!} \left( \text{val}(S \cup \{j\}) - \text{val}(S) \right)
$$

Good lord!

Let's break down this expression:

1.$\phi_j(\text{val})$: This is the **Shapley value** for feature $j$ in the function or model $\text{val}$. It represents the **contribution** of feature $j$ to the final output of the model. 

   - **val**: The value function (or payoff function in game theory), which measures the output or gain from a set of features (or players).
   - **j**: The index of the feature (or player) for which we are computing the Shapley value.

2. **$S \subseteq \{1, \dots, p\} \setminus \{j\}$**: 
   - This is a **subset** $S$ of all possible features (or players), except the feature $j$ whose contribution we're trying to calculate.
   - It represents all the possible coalitions of features (or players) without including feature $j$.
   - Essentially, we look at all subsets $S$ of features that do not contain $j$, and evaluate the effect of adding $j$ to each subset.

3. **$\frac{|S|! (p - |S| - 1)!}{p!}$**: 
   - This is a **weight** based on the size of the subset $S$. It ensures that each subset is considered in a balanced and fair way.
   - Here's a breakdown of this term:
     - $|S|!$: The factorial of the size of $S$, which accounts for the number of ways the features in $S$ could be arranged.
     - $(p - |S| - 1)!$: The factorial of the remaining features that are not in $S$ or $j$, which accounts for the number of ways the remaining features can be arranged.
     - $p!$: The factorial of the total number of features, ensuring the whole expression is normalized (so the contributions from all possible subsets sum up to the full model value).

4. **$\text{val}(S \cup \{j\}) - \text{val}(S)$**: 
   - This term represents the **marginal contribution** of feature $j$ to the subset $S$. 
   - $\text{val}(S \cup \{j\})$: The output of the value function when feature $j$ is included in the subset $S$.
   - $\text{val}(S)$: The output of the value function when feature $j$ is not included.
   - The difference $\text{val}(S \cup \{j\}) - \text{val}(S)$ shows how much the prediction (or game value) changes when feature $j$ is added to the subset $S$.


A good **example** that can increase the comprehention of the concept may be read below:

Imagine a room where guests decide on a matter, and the goal is to decide on the answer to a certain question. In the beginning, before anyone enters the room, there's just a default answer. Let's say it's something vague or neutral, like "We don't know yet."

Now, one by one, different people enter the room, each bringing their unique knowledge or perspective. These people are like the features in a machine learning model. Each time someone enters the room, they add value to the decision-making process. The final decision, when everyone is inside the room, is the cumulative result of all the contributions.

However, the order in which people enter the room matters for evaluating their contributions, as it may change the way their input is perceived or combined with others'. This is where the Shapley value idea comes in.

## Properties

SHAP values have several useful properties that make them effective for interpreting models:

1. Efficiency (Pareto Optimality): This property ensures that the total value (or payoff) of the game (or model prediction) is completely distributed among all players (or features). The sum of all individual Shapley values is equal to the total value generated by the coalition of all players.

$$
\sum_{j=1}^{p} \phi_j(\text{val}) = \text{val}(\{1, \dots, p\})
$$

2. Symmetry: If two players (or features) contribute the same value to every possible coalition, then their Shapley values should be equal. This ensures that no two features that are equally important (in every context) are treated differently.

$$
\text{If for all } S \subseteq \{1, \dots, p\} \setminus \{i, j\}: \quad \text{val}(S \cup \{i\}) = \text{val}(S \cup \{j\}) \\
\text{Then:} \quad \phi_i(\text{val}) = \phi_j(\text{val})
$$

3. Linearity (Additivity): The Shapley value is linear in the value function. If two games (or models) are combined, the Shapley value for each player (or feature) in the combined game is the sum of their Shapley values in each individual game.
$$
\phi_j(\text{val}_1 + \text{val}_2) = \phi_j(\text{val}_1) + \phi_j(\text{val}_2)
$$


4. Null Player (Dummy Player): If a player (or feature) does not contribute anything to any coalition, its Shapley value should be zero. This property ensures that features that add no value are not assigned any contribution.
$$
\text{If for all } S \subseteq \{1, \dots, p\}: \quad \text{val}(S \cup \{j\}) = \text{val}(S) \\
\text{Then:} \quad \phi_j(\text{val}) = 0
$$

5. Marginality: The Shapley value reflects the marginal contribution of each feature. The marginal contribution is how much the inclusion of a feature improves the prediction or value, averaged over all possible subsets of features.

$$
\text{Marginal contribution of } j \text{ to } S = \text{val}(S \cup \{j\}) - \text{val}(S)
$$

In summary:

1. Efficiency guarantees that the total output is fully distributed among the features.
2. Symmetry ensures that equally contributing features receive equal credit.
3. Linearity and Additivity allow for the Shapley value to work with combined or separated models.
4. Null Player ensures that uninformative features are assigned zero contribution.
5. Marginality highlights that each feature's contribution is measured fairly across all possible subsets.

These properties are what make the Shapley value a robust and mathematically sound method for explaining feature importance in machine learning models.

# Manually calculating SHAP values for a linear regression

What I've done up there is basically a lot of math to **represent** the **idea** behind the estimation of a shapley value. But the **real** math may vary depending on the situation. I'll provide a simple solution that is computationally expensive.

The idea here is quite simple: for every possible coalition generated on a model with 3 features (Intercept, B1 and B2), generate mean predicted values so that we can fill them inside the shap formula. (I'll implement a function to do this automatic later.)

A important note is that every contribution is computed in a instance level, which is the same as computing FOR EVERY "ROW".

In [3]:
import pandas as pd
import numpy as np

from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression

from itertools import chain, combinations 

X, y = make_regression(n_samples=100, n_features=3, noise=0.1, random_state=42)
df_X = pd.DataFrame(X, columns=['Feature_1', 'Feature_2', 'Feature_3'])
df_y = pd.DataFrame(y, columns=['Target'])
df = pd.concat([df_X, df_y], axis=1)

print(f"This is my mock up dataset:\n {df.head(10)}")

This is my mock up dataset:
    Feature_1  Feature_2  Feature_3      Target
0  -0.792521   0.504987  -0.114736   13.510026
1   0.280992  -0.208122  -0.622700  -18.777475
2   0.791032   1.402794  -0.909387  111.265809
3   0.625667  -1.070892  -0.857158  -77.989347
4  -0.342715  -0.161286  -0.802277  -35.951738
5   2.122156  -1.519370   1.032465  -35.839901
6  -1.401851   2.190456   0.586857  135.451953
7  -0.908024   1.465649  -1.412304   59.303696
8   0.259883  -1.236951   0.781823  -71.595854
9  -0.815810   0.341152  -0.077102    1.236046


In [4]:
# Just messing up here, how are the coefficients for this linear model using OLS
model = LinearRegression()
model.fit(df_X, df_y)
model.coef_

array([[28.2045949 , 75.05077568, 17.75449804]])

In [5]:
features = ["Feature_1", "Feature_2", "Feature_3"]

superset = [i for i in chain.from_iterable(combinations(features, r) for r in range(len(features)+1))]
print(f"The superset defined by 3 features is: {superset}")

The superset defined by 3 features is: [(), ('Feature_1',), ('Feature_2',), ('Feature_3',), ('Feature_1', 'Feature_2'), ('Feature_1', 'Feature_3'), ('Feature_2', 'Feature_3'), ('Feature_1', 'Feature_2', 'Feature_3')]


In [6]:
instance = df.loc[0, :]
for game in superset:
    assert len(instance.shape) == 1, 'Instance must be a 1D array'
    prediction = 0
    if len(game) == 0:
        prediction = y.mean()
    else:
        model = LinearRegression()
        model.fit(df[[feature for feature in game]].values, df["Target"])
        prediction = model.predict(instance.loc[[feature for feature in game]].values.reshape(1, -1))

    print(game, prediction)

() 4.962922789987591
('Feature_1',) [-8.26237738]
('Feature_2',) [35.33023419]
('Feature_3',) [5.37670622]
('Feature_1', 'Feature_2') [12.6247103]
('Feature_1', 'Feature_3') [-8.14604212]
('Feature_2', 'Feature_3') [37.20101816]
('Feature_1', 'Feature_2', 'Feature_3') [13.52236521]


For a single instance, we shall have the following coalisions:
$$
() = 4.96 \newline
(Feature_1) = -8.26 \newline
(Feature_2) = 35.33 \newline
(Feature_3) = 5.37 \newline
(Feature_1, Feature_2) = 12.62 \newline
(Feature_1, Feature_3) = -8.14 \newline
(Feature_2, Feature_3) = 37.20 \newline
(Feature_1, Feature_2, Feature_3) = 13.52
$$

Note that *a model with no features predicts E[y].*

Taking the big equation, I'll start solving it manually.
$$
\phi_j(\text{val}) = \sum_{S \subseteq \{1,\dots,p\} \setminus \{j\}} \frac{|S|! (p - |S| - 1)!}{p!} \left( \text{val}(S \cup \{j\}) - \text{val}(S) \right)
$$

For the intercept, we must exclude all coalisions that the intercept takes part and solve the summation. In this case we have 4 coalisions that intercept does not participate.
$$
\phi_1(\text{val}) = \frac{1}{3!}( \newline
                                0! (3 - 0 - 1)! \left[ \text{val}(\{Feature_1\}) - \text{val}(\empty) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_1, Feature_2\}) - \text{val}(Feature_2) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_1, Feature_3\}) - \text{val}(Feature_3) \right] + \newline
                                2! (3 - 2 - 1)! \left[ \text{val}(\{Feature_1, Feature_2, Feature_3\}) - \text{val}(Feature_2, Feature_3) \right])
                                
$$

Using the values we previously calculated:

$$
\phi_1(\text{val}) = \frac{1}{6}(0! (2)! [-8.26 - 4.96] + 1! (1)! [ 12.62 - 35.33] + 1! (1)! [ -8.14 - 5.37 ] + 2! (0)! [ 13.52 - 37.20 ])                    
$$

$$
\phi_1(\text{val}) = \frac{1}{6}(1 \cdot 2 \cdot -13.22 + 1 \cdot 1 \cdot -22.71 + 1 \cdot 1 \cdot -13.51 + 2 \cdot 1 \cdot -23.68) = -18.336 
$$

Doing the same for the others:
$$
\phi_2(\text{val}) = \frac{1}{3!}( \newline
                                0! (3 - 0 - 1)! \left[ \text{val}(\{Feature_2\}) - \text{val}(\empty) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_1, Feature_2\}) - \text{val}(Feature_1) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_2, Feature_3\}) - \text{val}(Feature_3) \right] + \newline
                                2! (3 - 2 - 1)! \left[ \text{val}(\{Feature_1, Feature_2, Feature_3\}) - \text{val}(Feature_1, Feature_3) \right])
                                
$$
$$
\phi_2(\text{val}) = \frac{1}{6}(0! (2)! [35.33 - 4.96] + 1! (1)! [ 12.62 + 8.26] + 1! (1)! [ 37.20 - 5.37 ] + 2! (0)! [ 13.52 + 8.14 ])                    
$$
$$
\phi_2(\text{val}) = \frac{1}{6}(1 \cdot 2 \cdot 30.37 + 1 \cdot 1 \cdot 20.88 + 1 \cdot 1 \cdot 31.83 + 2 \cdot 1 \cdot 21.66) = 26.13
$$

$$
\phi_3(\text{val}) = \frac{1}{3!}( \newline
                                0! (3 - 0 - 1)! \left[ \text{val}(\{Feature_3\}) - \text{val}(\empty) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_1, Feature_3\}) - \text{val}(Feature_1) \right] + \newline 
                                1! (3 - 1 - 1)! \left[ \text{val}(\{Feature_2, Feature_3\}) - \text{val}(Feature_2) \right] + \newline
                                2! (3 - 2 - 1)! \left[ \text{val}(\{Feature_1, Feature_2, Feature_3\}) - \text{val}(Feature_1, Feature_2) \right])
                                
$$
$$
\phi_3(\text{val}) = \frac{1}{6}(0! (2)! [5.37 - 4.96] + 1! (1)! [ -8.14 + 8.26] + 1! (1)! [ 37.20 - 35.33 ] + 2! (0)! [ 13.52 - 12.62])                    
$$
$$
\phi_3(\text{val}) = \frac{1}{6}(1 \cdot 2 \cdot 0.41 + 1 \cdot 1 \cdot 0.12 + 1 \cdot 1 \cdot 1.87 + 2 \cdot 1 \cdot 0.90) = 0.76
$$

**Note that all of those calculations were done for a single instance of our original dataset. This is why computing shapvalues is quite an expensive feat!**

$$
(\phi_1(\text{val}), \phi_2(\text{val}), \phi_3(\text{val})) = (-18.336, 26.13, 0.76)
$$

Below, I'll create a simple function to generate shapvalues for every single instance of our 100 row dataset. The "time complexity" for a optimal solution of this problem would be something near O(2^p) with p as features.

In [19]:
# I'll base this implementation on the one here https://randomrealizations.com/posts/shap-from-scratch/#what-is-shap
from math import factorial

class ShapExplainerByMyself:
    def __init__(self):
        pass

    def explain(self, X, y, feature_of_interest = 1):
        self.n_features = X.shape[1]
        self.y_mean = y.mean()

        self.X = X
        self.y = y

        self.model_cache = {}
        return self.compute_single_shap_value(feature_of_interest)
        
    def get_all_other_feature_subsets(self, feature_of_interest):
        all_other_features = [j for j in range(self.n_features) if j != feature_of_interest]
        return chain.from_iterable(combinations(all_other_features, r) for r in range(len(all_other_features)+1)) # THIS IS |S|
    
    def subset_model(self, feature_subset, instance):
        # This is the same as calculating the VALUE of given subset fitting a specific model for the features of given subset
        assert len(instance.shape) == 1, 'Instance must be a 1D array'
        if len(feature_subset) == 0:
            return self.y_mean # a model with no features predicts E[y]
        
        if tuple(feature_subset) in self.model_cache:
            model = self.model_cache[tuple(feature_subset)]
        else:
            X_subset = self.X.take(feature_subset, axis=1)
            model = LinearRegression()
            model.fit(X_subset, self.y)
            self.model_cache[tuple(feature_subset)] = model

        return model.predict(instance.take(feature_subset).reshape(1, -1))[0]
    
    def permutation_factor(self, n_subset):
        # This is |S!| * (j - |S| - 1)! / j!
        return factorial(n_subset) * factorial(self.n_features - n_subset - 1) / factorial(self.n_features)
    
    def compute_single_shap_value(self, feature_of_interest, instance = X[0, :]):
        shap_value = 0
        for subset in self.get_all_other_feature_subsets(feature_of_interest):
            n_subset = len(subset)
            prediction_without_feature = self.subset_model(
                subset,
                instance
            )
            prediction_with_feature = self.subset_model(
                subset + (feature_of_interest,),
                instance
            )
            factor = self.permutation_factor(n_subset)
            shap_value += factor * (prediction_with_feature - prediction_without_feature)
        return shap_value

In [20]:
# Let's check if results are consistent:
explainer = ShapExplainerByMyself()
print(explainer.explain(X, y, 0), explainer.explain(X, y, 1), explainer.explain(X, y, 2))

-18.3393630786159 26.130472847745505 0.7683326528153469


They are! Now I'll implement a logic to compute for every instance in the dataset.

In [26]:
# I'll change my class a little bit to look more like the shap explainer class
class ShapExplainerByMyself:
    def __init__(self, X, y):
        self.X = X
        self.y = y
        self.n_features = X.shape[1]
        self.y_mean = y.mean()
        self.model_cache = {}

    def explain(self):
        output = pd.DataFrame().reindex_like(self.X) # I'll use pandas b/c of compatibility but numpy would be best here.
        for row in output.index:
            for feature in range(self.n_features):
                output.iloc[row, feature] = self.compute_single_shap_value(feature_of_interest= feature, instance = X[row, :])
        return output
        
    def get_all_other_feature_subsets(self, feature_of_interest):
        all_other_features = [j for j in range(self.n_features) if j != feature_of_interest]
        return chain.from_iterable(combinations(all_other_features, r) for r in range(len(all_other_features)+1)) # THIS IS |S|
    
    def subset_model(self, feature_subset, instance):
        # This is the same as calculating the VALUE of given subset fitting a specific model for the features of given subset
        assert len(instance.shape) == 1, 'Instance must be a 1D array'
        if len(feature_subset) == 0:
            return self.y_mean # a model with no features predicts E[y]
        
        if tuple(feature_subset) in self.model_cache:
            model = self.model_cache[tuple(feature_subset)]
        else:
            X_subset = self.X.take(feature_subset, axis=1)
            model = LinearRegression()
            model.fit(X_subset, self.y)
            self.model_cache[tuple(feature_subset)] = model

        return model.predict(instance.take(feature_subset).reshape(1, -1))[0]
    
    def permutation_factor(self, n_subset):
        # This is |S!| * (j - |S| - 1)! / j!
        return factorial(n_subset) * factorial(self.n_features - n_subset - 1) / factorial(self.n_features)
    
    def compute_single_shap_value(self, feature_of_interest, instance):
        shap_value = 0
        for subset in self.get_all_other_feature_subsets(feature_of_interest):
            n_subset = len(subset)
            prediction_without_feature = self.subset_model(
                subset,
                instance
            )
            prediction_with_feature = self.subset_model(
                subset + (feature_of_interest,),
                instance
            )
            factor = self.permutation_factor(n_subset)
            shap_value += factor * (prediction_with_feature - prediction_without_feature)
        return shap_value

In [38]:
import warnings
warnings.filterwarnings("ignore")

explainer = ShapExplainerByMyself(df_X, y)
shap_values = explainer.explain()

# SHAP feature-wise values (averaged over all data points)
shap_feature_importance = shap_values.values.mean(axis=0)

# Print feature-wise SHAP values
print(f"SHAP values for Feature 1: {shap_feature_importance[0]}")
print(f"SHAP values for Feature 2: {shap_feature_importance[1]}")
print(f"SHAP values for Feature 3: {shap_feature_importance[2]}")


SHAP values for Feature 1: -1.8474111129762605e-15
SHAP values for Feature 2: -4.263256414560601e-16
SHAP values for Feature 3: 4.1744385725905884e-16


In [41]:
import shap

# Create a SHAP explainer for the linear regression model
model = LinearRegression()
model.fit(df_X, df_y)
explainer = shap.LinearExplainer(model, df_X)

# Calculate SHAP values for the dataset
shap_values = explainer(df_X)

# SHAP feature-wise values (averaged over all data points)
shap_feature_importance = shap_values.values.mean(axis=0)

# Print feature-wise SHAP values
print(f"SHAP values for Feature 1: {shap_feature_importance[0]}")
print(f"SHAP values for Feature 2: {shap_feature_importance[1]}")
print(f"SHAP values for Feature 3: {shap_feature_importance[2]}")

SHAP values for Feature 1: -3.552713678800501e-16
SHAP values for Feature 2: -3.552713678800501e-15
SHAP values for Feature 3: 4.796163466380677e-16


My implementation calculated SHAP values without approximation, whereas the SHAP package uses an approximation to make the computation feasible for large datasets and complex models. The fact that the results are so close, despite the SHAP package using kernel-based approximations, suggests that the package is highly accurate.

The small discrepancies between my SHAP values and those of the SHAP package are most likely due to:
   - Kernel-based approximations used by the SHAP package.
   - The inherent limitations of floating-point precision in computers.
   - The fact that I calculated SHAP values without approximation, whereas the SHAP package balances accuracy and computational efficiency.

   These differences are negligible and do not affect the overall interpretation of feature importance in the model.

# References

I saw [this](https://youtu.be/fbrVvMU8T6o?si=G52c-L4T2uWGaGS8) youtube video to grasp on how to make the calculation more clearly and read the articles below:

https://christophm.github.io/interpretable-ml-book/shap.html

https://randomrealizations.com/posts/shap-from-scratch/#what-is-shap